### PR TITLE
TypeScript version

### PR DESCRIPTION
**Pull request by piliugin-anton**

### CODE REVIEW

**Comment by Alhadis** [comment]
OMG, you converted the entire project to TypeScript. I thought you were referring to fixing the type declarations (`pan-and-zoom.d.ts`).

Tell you what: you clearly care about this project more than I do, and I'd be happy to transfer ownership of it to you. This was written to support a [project](https://github.com/Alhadis/Atom-LiveView) that I later axed, and I've very little motivation to continue working on it.

I adamantly refuse to work with Webpack/Babel/TypeScript/whatever-else-is-flavour-of-the-month-at-the-moment. But I'd also feel like a complete cunt if I rejected your hard work, so…

**Comment by piliugin-anton** [comment]
I didn't worked with TypeScript because I was thinking it's just a trendy thing, but it looks like it's there forever and employers nowadays expecting from potential candidates to be familiar with it. You can ignore this PR, because I'm still thinking about removing a string arguments support to be able to use += operator

**Comment by Alhadis** [comment]
> but [TypeScript] looks like it's there forever and employers nowadays expecting from potential candidates to be familiar with it.

😉 Aye, but I quit the industry years ago. Now I work with web technologies purely as a hobbyist, which gives me the luxury of being an obstinate purist prick when it comes to choosing what I work with.

> because I'm still thinking about removing a string arguments support to be able to use += operator

Mate, go nuts. Once I transfer ownership of this to you, you can do whatever you want to it.

**Comment by piliugin-anton** [comment]
@Alhadis Could you, please, give me a guide/links on matrices (of why they are multiplied and not summed, for example) and how they are applied to a canvas/image (what exactly happens step-by-step)?

**Comment by Alhadis** [comment]
@piliugin-anton I'm in the middle of writing something to help explain matrices and how they're used. In the mean time, you'll need to integrate a small change I made that simplifies conversion to CSS/canvas-compatible matrix arrays (which obviates the `toCanvasMatrix()` that you've added).

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+  extends: ['@alhadis/eslint-config'],
+  parser: '@typescript-eslint/parser',
+  plugins: ['@typescript-eslint'],
+};
```
Comment (Alhadis) [inline .eslintrc.js]: You may wish to configure your editor to use [EditorConfig](https://editorconfig.org/) which should help ensure consistent formatting when editing/creating files. 😉

File: config/tsconfig.cjs.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+    "extends": "../tsconfig",
+    "compilerOptions": {
+      "module": "commonjs" /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */,
```
Comment (Alhadis) [inline config/tsconfig.cjs.json:4]: If you'd like these comments to stop being highlighted as errors, you should mark these files as JSONC. There are two ways to do that:

1.	Give every `.json` file that uses JSONC the `.jsonc` extension
2.	Mark these files in `.gitattributes` as JSONC, not JSON:
	~~~gitattributes
	config/*.json linguist-language=JSONC
	~~~
	More about that [here](https://github.com/github/linguist/blob/master/docs/overrides.md).

**Comment by piliugin-anton** [comment]
@Alhadis Added iterator, updated README and CHANGELOG

**Comment by Alhadis** [comment]
> what exactly happens step-by-step?

The [`applyTransform()`](https://github.com/Alhadis/PanAndZoom/blob/c4e650f1e4e7dbc5772ed4bb9fde9573fdb1bbe7/pan-and-zoom.mjs#L497-L508) method is the surface-level mechanism for transforming data. Essentially, it takes as input an array of points (each expressed as an array of two numbers, X and Y). Let's start with something super simple:

~~~js
import PanAndZoom from "./pan-and-zoom.mjs";
import assert from "assert";

// Array of points making up the corners of a unitless 100×100 square
let points = [
	[0,   0],   // Top-left corner
	[100, 0],   // Top-right corner
	[100, 100], // Bottom-right corner
	[0,   100], // Bottom-left corner
];

const paz = new PanAndZoom();

// Move each point along its “X” axis by 50
paz.panX = 50;
assert.deepStrictEqual(
	paz.applyTransform(points),
	[
		[50,  0],
		[150, 0],
		[150, 100],
		[50,  100],
	],
);

// Move each point along its "Y" axis by 25
paz.panY = 25;
assert.deepStrictEqual(
	paz.applyTransform(points),
	[
		[50,  25],
		[150, 25],
		[150, 125],
		[50,  125],
	],
);
~~~

Now, for the second example, and to answer your other question:

> why they are multiplied and not summed

Translation is the simplest transformation possible with a matrix: it's literally just adding values together. But transform matrices immediately become a lot more useful when you mix other transformation effects: [scaling/zooming](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/scale), [rotation](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/rotate), [shearing/skewing](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/skew), etc. 

Continuing on from the previous example, let's try zooming in +200% with the effect centred on an arbitrary position:

~~~js
// Zoom +150% centred on [120, 80]
paz.zoom    = 1.5;
paz.originX = 120;
paz.originY = 80;
assert.deepStrictEqual(
	paz.applyTransform(points),
	[
		[-10, -15],
		[140, -15],
		[140, 135],
		[-10, 135],
	]
);
~~~

Now, you *could* loop through each point, offset each one's X and Y ordinate by the current pan settings, and then multiply the result by the current zoom amount. The problem is that's not performant when dealing with huge lists of spatial data (think: pixels or detailed vector images) whereas a transform matrix is generally much more efficient.

> how they are applied to a canvas/image

That's where you're in luck. The HTML Canvas API provides [`getTransform()`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getTransform)/[`setTransform()`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setTransform) methods out-of-the-box. I'd give you an example, but I'm currently having trouble trying to figure out why `getTransform([...paz])` or `getTransform(...paz)` isn't working… :thinking:

**Comment by piliugin-anton** [comment]
Thanks for explanation, I will read the second part few times to better get the idea of how it works

> I'd give you an example, but I'm currently having trouble trying to figure out why `getTransform([...paz])` or `getTransform(...paz)` isn't working… thinking

I've fixed your iterator, because the order of parameters wasn't correct, take a look at my last commit

**Comment by Alhadis** [comment]
> I've fixed your iterator, because the order of parameters wasn't correct

🤦‍♂️ Ugh, that was a hasty [copy-and-paste fail](https://github.com/Alhadis/PanAndZoom/commit/8f47e95a020a0f6dcc3b1918375732af48ca3cad#diff-d652a1b949c388e21a77088c46c4d301031355815938d1bb76cc63115e11141bL592-L593). It's late over here in Melbourne and I need sleep… 😅

File: CHANGELOG.md [hunk 3]
```diff
@@ -19,3 +19,12 @@ This project honours [Semantic Versioning](http://semver.org/).
...
 [v1.0.0]: https://github.com/Alhadis/PanAndZoom/releases/tag/v1.0.0
+
+
+
+[v1.0.1]
+------------------------------------------------------------------------
+**September 4th, 2022**  
+Rewritten to TypeScript. Contributed by [Anton Piliugin](https://github.com/piliugin-anton)
+
+[v1.0.1]: https://github.com/Alhadis/PanAndZoom/releases/tag/v1.0.1
```
Comment (Alhadis) [inline CHANGELOG.md]: This should go under `[Unpublished]` instead. 😉

Only upon release is `[Unpublished]` replaced with the name of the new tag, a date added under the divider-line, and a link to the release added:

~~~diff
-[Unpublished]: ../../compare/v1.0.0...HEAD
+[Unpublished]: ../../compare/v1.0.1...HEAD
 
-[Unpublished]
+[v1.0.1]
 ------------------------------------------------------------------------
+**September 5th, 2022**
 * **Fixed:** Breakage when loading non-ESM version in browsers
 * **Fixed:** Values not clamped or typecast if set during initialisation
+
+[v1.0.1]: https://github.com/Alhadis/PanAndZoom/releases/tag/v1.0.1
~~~

That's just how I prefer to format change-logs (with newer releases appearing at the top, and older releases at the bottom). You're free to use whichever format you like most once/if this repository's traded hands. 😉

**Review by Alhadis — CHANGES_REQUESTED** [review]

File: .eslintrc.js (hunk 1 above)
Comment (piliugin-anton) [reply .eslintrc.js]: I left the settings, just because to keep your settings that are suitable for you. I personally use 'recommended' config

Comment (piliugin-anton) [reply .eslintrc.js]: Is this EditorConfig does the same ESLint does?

Comment (Alhadis) [reply .eslintrc.js]: ESLint's recommended settings only address things that are objectively problematic: misspelt or unused variables, undefined references, obsolete and confusing features of JS nobody uses any more… etc. It *does not* include stylistic and formatting decisions, which are purely preferential in nature.

EditorConfig focusses on such stylistic changes, and it exists to help maintain coding styles between projects, which inevitably differ between authors.

File: CHANGELOG.md [hunk 4]
```diff
@@ -3,13 +3,16 @@ Change Log
...
-------------------------------------------------------------------------
-* **Fixed:** Breakage when loading non-ESM version in browsers
-* **Fixed:** Values not clamped or typecast if set during initialisation
+[v1.0.1]
+ ------------------------------------------------------------------------
+**September 5th, 2022**
+ * **Fixed:** Breakage when loading non-ESM version in browsers
+ * **Fixed:** Values not clamped or typecast if set during initialisation
+
+[v1.0.1]: https://github.com/Alhadis/PanAndZoom/releases/tag/v1.0.1
```
Comment (Alhadis) [inline CHANGELOG.md]: Oops, I accidentally counted too many spaces when writing that diff before… 😓 Also, the date and version number is only added right before the release is about to be published. Otherwise, there's a risk that a delay or new problem will come up, and you'll overlook the wrong date and version in the config (I've made that mistake quite a few times 😉).

```suggestion
[v1.0.1]
------------------------------------------------------------------------
**September 5th, 2022**
* **Fixed:** Breakage when loading non-ESM version in browsers
* **Fixed:** Values not clamped or typecast if set during initialisation

[v1.0.1]: https://github.com/Alhadis/PanAndZoom/releases/tag/v1.0.1
```

File: test/pan-and-zoom.spec.ts [hunk 5]
```diff
@@ -1,6 +1,5 @@
-"use strict";
-
-const PanAndZoom = require("../pan-and-zoom.js");
+/* eslint-disable no-undef */
```
Comment (Alhadis) [inline test/pan-and-zoom.spec.ts]: Why was this added?

File: test/pan-and-zoom.spec.ts [hunk 6]
```diff
@@ -123,12 +122,12 @@ describe("PanAndZoom", () => {
...
-				paz.minPanX = undefined; expect(paz.minPanX).to.be.a("number").that.equals(0);
-				paz.minPanY = null;      expect(paz.minPanY).to.be.a("number").that.equals(0);
-				paz.maxPanX = "Nope";    expect(paz.maxPanX).to.be.a("number").that.equals(0);
-				paz.maxPanY = NaN;       expect(paz.maxPanY).to.be.a("number").that.equals(0);
+				paz.panX    = [] as any;        expect(paz.panX).to.be.a("number").that.equals(0);
+				paz.panY    = {} as any;        expect(paz.panY).to.be.a("number").that.equals(0);
+				paz.minPanX = undefined as any; expect(paz.minPanX).to.be.a("number").that.equals(0);
+				paz.minPanY = null as any;      expect(paz.minPanY).to.be.a("number").that.equals(0);
+				paz.maxPanX = "Nope";    		expect(paz.maxPanX).to.be.a("number").that.equals(0);
+				paz.maxPanY = NaN;       		expect(paz.maxPanY).to.be.a("number").that.equals(0);
```
Comment (Alhadis) [inline test/pan-and-zoom.spec.ts]: Tabs should only be [used for indentation](https://github.com/Alhadis/Coding-Style#1-use-tabs-for-indentation), not for alignment. 😉 That is:

1. Tabs at the start of a line
2. Spaces after non-whitespace characters
3. Never let tabs and spaces touch

Comment (Alhadis) [reply test/pan-and-zoom.spec.ts]:
```suggestion
				paz.maxPanX = "Nope";           expect(paz.maxPanX).to.be.a("number").that.equals(0);
				paz.maxPanY = NaN;              expect(paz.maxPanY).to.be.a("number").that.equals(0);
```

File: test/pan-and-zoom.spec.ts [hunk 7]
```diff
@@ -293,21 +292,21 @@ describe("PanAndZoom", () => {
 				minZoom: -2,
 				maxZoom: +4,
 			});
-			paz.zoom    = [];        expect(paz.zoom).to.be.a("number").that.equals(0);
-			paz.minZoom = {};        expect(paz.minZoom).to.be.a("number").that.equals(0);
-			paz.maxZoom = "Invalid"; expect(paz.maxZoom).to.be.a("number").that.equals(0);
+			paz.zoom    = [] as any;    expect(paz.zoom).to.be.a("number").that.equals(0);
+			paz.minZoom = {} as any;    expect(paz.minZoom).to.be.a("number").that.equals(0);
+			paz.maxZoom = "Invalid"; 	expect(paz.maxZoom).to.be.a("number").that.equals(0);
```
Comment (Alhadis) [inline test/pan-and-zoom.spec.ts]:
```suggestion
			paz.maxZoom = "Invalid";    expect(paz.maxZoom).to.be.a("number").that.equals(0);
```

File: test/pan-and-zoom.spec.ts [hunk 8]
```diff
@@ -432,12 +431,12 @@ describe("PanAndZoom", () => {
...
-				paz.minOriginX = undefined; expect(paz.minOriginX).to.be.a("number").that.equals(0);
-				paz.minOriginY = null;      expect(paz.minOriginY).to.be.a("number").that.equals(0);
-				paz.maxOriginX = "Nope";    expect(paz.maxOriginX).to.be.a("number").that.equals(0);
-				paz.maxOriginY = NaN;       expect(paz.maxOriginY).to.be.a("number").that.equals(0);
+				paz.originX    = [] as any;        expect(paz.originX).to.be.a("number").that.equals(0);
+				paz.originY    = {} as any;        expect(paz.originY).to.be.a("number").that.equals(0);
+				paz.minOriginX = undefined as any; expect(paz.minOriginX).to.be.a("number").that.equals(0);
+				paz.minOriginY = null as any;      expect(paz.minOriginY).to.be.a("number").that.equals(0);
+				paz.maxOriginX = "Nope";    	   expect(paz.maxOriginX).to.be.a("number").that.equals(0);
+				paz.maxOriginY = NaN;       	   expect(paz.maxOriginY).to.be.a("number").that.equals(0);
```
Comment (Alhadis) [inline test/pan-and-zoom.spec.ts]:
```suggestion
				paz.maxOriginX = "Nope";           expect(paz.maxOriginX).to.be.a("number").that.equals(0);
				paz.maxOriginY = NaN;              expect(paz.maxOriginY).to.be.a("number").that.equals(0);
```

**Review by Alhadis — CHANGES_REQUESTED** [review]

File: test/pan-and-zoom.spec.ts (hunk 5 above)
Comment (piliugin-anton) [reply test/pan-and-zoom.spec.ts]: ESLint thinks that 'expect' is not defined

Comment (Alhadis) [reply test/pan-and-zoom.spec.ts]: Oh, right. You've added an `.eslintrc.js` file, which shadows the [`"eslintConfig"`](https://github.com/Alhadis/PanAndZoom/blob/00d5cdc0ca2ae7d99d135d1441f7fb6cf18646d6/package.json#L26-L35) field in `package.json`. You should merge the contents of that entry with your `.eslintrc.js`: it defines `expect` as a global variable for ESLint so it won't complain about it.

File: .eslintrc.js [hunk 9]
```diff
@@ -2,4 +2,18 @@ module.exports = {
   extends: ['@alhadis/eslint-config'],
```
Comment (Alhadis) [inline .eslintrc.js]: **Tip:** ESLint allows you to shorten the name of [shareable configs](https://eslint.org/docs/latest/developer-guide/shareable-configs#using-a-shareable-config) if they match one of these formats:

```
"@user/eslint-config" -> "@user"
"eslint-config-user"  -> "user"
```

```suggestion
  extends: '@alhadis',
```

File: .eslintrc.js [hunk 10]
```diff
@@ -2,4 +2,18 @@ module.exports = {
...
+  overrides: [
+    {
+      "files": [
+        "test/*"
+      ],
+      "globals": {
+        "expect": true
+      }
+    }
+  ]
```
Comment (Alhadis) [inline .eslintrc.js]: Just curious, what text-editor do you use? EditorConfig support should've formatted that appropriately… 🤔

Comment (piliugin-anton) [reply .eslintrc.js]: I didn't added a proper extension yet. I'm using VSCode.

Comment (Alhadis) [reply .eslintrc.js]: I thought as much. VSCode claims to have [EditorConfig support built-in](https://docs.microsoft.com/en-us/visualstudio/releasenotes/vs2017-relnotes-v15.0#coding-convention-support-through-editorconfig), it just needs configuration.

Comment (piliugin-anton) [reply .eslintrc.js]: I think this page describes Visual Studio features. Visual Studio Code is a lightweight version of Visual Studio and I need to install an extension. Stupid naming for sure =)

Comment (piliugin-anton) [reply .eslintrc.js]: Btw I just installed it (I mean extension for EditorConfig).

Comment (Alhadis) [reply .eslintrc.js]: OH. Uh, okay, that makes things a bit more interesting. 🤣 I've never used Visual Studio, as you can probably tell (I'm from Unix-land 😉).

You might want to [check this out instead then](https://docs.microsoft.com/en-us/visualstudio/releasenotes/vs2017-mac-relnotes-v7.5#release-highlights).

Comment (piliugin-anton) [reply .eslintrc.js]: > OH. Uh, okay, that makes things a bit more interesting. rofl I've never used Visual Studio, as you can probably tell (I'm from Unix-land wink).

Yes, it looks like you are a system programmer (C and/or C++ probably) because each and every JS project repo you own contains Makefile =)

Comment (piliugin-anton) [reply .eslintrc.js]: You may be surprised, but VSCode is built using Electron (https://www.electronjs.org/) which is based on Node.js

Comment (Alhadis) [reply .eslintrc.js]: > Yes, it looks like you are a system programmer

HAH! Nope, wrong. 😉 I've been working with web technologies since... eh, 2004-2005, and most of my projects are actually pretty high-level stuff. C/C++'s pointer syntax scares the shit out of me. 😁

Nope, I use Makefiles because they're they're clean and declarative, and encourage you to leverage existing system commands (of which there are plenty) instead of downloading a crapload of tools needed to execute an overly-complex workflow. I prefer that complexity stay in the codebase I'm trying to build. 😉

FYI, you can run Make and many other Unix tools on Windows by installing [GOW](https://github.com/bmatzelle/gow/wiki); and you'll only need to do so once, as it's a very unintuitive installation.

Comment (piliugin-anton) [reply .eslintrc.js]: > HAH! Nope, wrong. wink I've been working with web technologies since... eh, 2004-2005, and most of my projects are actually pretty high-level stuff. C/C++'s pointer syntax scares the shit out of me. grin

I just started to learn C++, it's useful for 2 reasons:
 - Node.js C++ bindings
 - WASM (Using existing C++ codebase you can make a WASM file and run in browser environment, see: https://emscripten.org/)

> Nope, I use Makefiles because they're they're clean and declarative, and encourage you to leverage existing system commands (of which there are plenty) instead of downloading a crapload of tools needed to execute an overly-complex workflow. I prefer that complexity stay in the codebase I'm trying to build. wink

If you already using NPM for your project, then it might be easier to just add a scripts in package.json

> FYI, you can run Make and many other Unix tools on Windows by installing [GOW](https://github.com/bmatzelle/gow/wiki); and you'll only need to do so once, as it's a very unintuitive installation.

I don't use Windows, because it's crap IMHO, I'm okay with Ubuntu Linux for now

Comment (Alhadis) [reply .eslintrc.js]: Wait, Visual Studio runs on Linux? 🤔 The more you know…

> If you already using NPM for your project

I'm not using NPM by choice. In fact, it's pretty much only used for installing/updating ESLint, C8, and Mocha/Chai for linting and running tests. And I'm trying to move away from the latter two.

> I just started to learn C++

I'd learn Rust instead. 😉 It's especially good at working with WebAssembly.

Comment (piliugin-anton) [reply .eslintrc.js]: > Wait, Visual Studio runs on Linux? thinking The more you know…

Yes, from the Electron link above you may find that it's cross-platform. It's not that hard to bundle Node.js + Chrome browser + Electron itself

Comment (Alhadis) [reply .eslintrc.js]: > It's not that hard to bundle Node.js + Chrome browser + Electron itself

It is on BSD systems. 😜 Shit, it took **years** to port Electron to OpenBSD…

Comment (piliugin-anton) [reply .eslintrc.js]: > It is on BSD systems. stuck_out_tongue_winking_eye Shit, it took **years** to port Electron to OpenBSD…

Here is a link on build (last message): https://github.com/marktext/marktext/issues/2655#issuecomment-1003303073

**Comment by Alhadis** [comment]
~~~
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To github.com:piliugin-anton/PanAndZoom.git
 ! [remote rejected] HEAD -> typescript (permission denied)
error: failed to push some refs to 'github.com:piliugin-anton/PanAndZoom.git'
~~~

@piliugin-anton Could you make your fork writeable by maintainers, please? I've resolved the merge conflicts with `master` locally.

**Review by Alhadis — APPROVED** [review]

File: .eslintrc.js (hunk 10 above)
Comment (Alhadis) [reply .eslintrc.js]: LOL, yeah, we pretty much cribbed FreeBSD's port and waited on _them_ to do all the hard work… 😁

**Comment by piliugin-anton** [comment]
Screenshot from 2022-09-05 02-29-36
I think this PR already have this option checked
Or it's somewhere else?

**Comment by Alhadis** [comment]
Ah, there we go. I think GitHub was complaining because I accidentally tried creating a new branch on your fork instead (word of advice: please always fork to a topic branch… it's a pain when resolving two `master` branches locally 😩).

**Comment by piliugin-anton** [comment]
> Ah, there we go. I think GitHub was complaining because I accidentally tried creating a new branch on your fork instead (word of advice: please always fork to a topic branch… it's a pain when resolving two `master` branches locally weary).

What do you mean by "fork to a topic branch"?

**Comment by Alhadis** [comment]
> What do you mean by "fork to a topic branch"?

When you create the fork of a repository you plan on submitting a pull-request to, it helps to create a new branch on your fork and do all your work there. Working directly on your fork's copy of `master` is still possible, but it complicates matters for maintainers who want to check your fork out locally (to verify that it works, etc).

Which reminds me, there appears to be versioning conflicts with the package's dependencies:

<details>

~~~
λ PanAndZoom (HEAD detached at fork/master): npm install
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: pan-and-zoom@v1.0.1
npm ERR! Found: eslint@5.16.0
npm ERR! node_modules/eslint
npm ERR!   dev eslint@"^5.16.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer eslint@"^6.0.0 || ^7.0.0 || ^8.0.0" from @typescript-eslint/parser@5.36.1
npm ERR! node_modules/@typescript-eslint/parser
npm ERR!   dev @typescript-eslint/parser@"^5.36.1" from the root project
npm ERR!   peer @typescript-eslint/parser@"^5.0.0" from @typescript-eslint/eslint-plugin@5.36.1
npm ERR!   node_modules/@typescript-eslint/eslint-plugin
npm ERR!     dev @typescript-eslint/eslint-plugin@"^5.36.1" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/Alhadis/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/Alhadis/.npm/_logs/2022-09-04T21_45_10_965Z-debug-0.log
~~~

</details>

File: .eslintrc.js (hunk 10 above)
Comment (piliugin-anton) [reply .eslintrc.js]: The version is quite old (8.2.0) there, while current stable release is [Electron 20.0.0](https://www.electronjs.org/releases/stable#20.0.0)

Comment (Alhadis) [reply .eslintrc.js]: Yep. That'd be because it's a nightmare to build, and IIRC, that version was only successfully ported by building against the Chromium port (as opposed to Electron's own version of Chrome).

I can't blame anybody for putting off wanting to update that pigsty of a codebase. 😫

**Comment by piliugin-anton** [comment]
> When you create the fork of a repository you plan on submitting a pull-request to, it helps to create a new branch on your fork and do all your work there. Working directly on your fork's copy of `master` is still possible, but it complicates matters for maintainers who want to check your fork out locally (to verify that it works, etc).

Okay, so next time I fork a project I should name it differently?

Also I just noticed that I forgot to remove 'ts-transform-default-export' from dependencies, I was trying to use it in order to make default imports work, but failed

**Comment by piliugin-anton** [comment]
> λ PanAndZoom (HEAD detached at fork/master): npm install
> npm ERR! code ERESOLVE
> npm ERR! ERESOLVE unable to resolve dependency tree
> npm ERR! 
> npm ERR! While resolving: pan-and-zoom@v1.0.1
> npm ERR! Found: eslint@5.16.0
> npm ERR! node_modules/eslint
> npm ERR!   dev eslint@"^5.16.0" from the root project
> npm ERR! 
> npm ERR! Could not resolve dependency:
> npm ERR! peer eslint@"^6.0.0 || ^7.0.0 || ^8.0.0" from @typescript-eslint/parser@5.36.1
> npm ERR! node_modules/@typescript-eslint/parser
> npm ERR!   dev @typescript-eslint/parser@"^5.36.1" from the root project
> npm ERR!   peer @typescript-eslint/parser@"^5.0.0" from @typescript-eslint/eslint-plugin@5.36.1
> npm ERR!   node_modules/@typescript-eslint/eslint-plugin
> npm ERR!     dev @typescript-eslint/eslint-plugin@"^5.36.1" from the root project
> npm ERR! 
> npm ERR! Fix the upstream dependency conflict, or retry
> npm ERR! this command with --force, or --legacy-peer-deps
> npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
> npm ERR! 
> npm ERR! See /Users/Alhadis/.npm/eresolve-report.txt for a full report.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/Alhadis/.npm/_logs/2022-09-04T21_45_10_965Z-debug-0.log
> ```

I could try to update the ESLint version

**Comment by Alhadis** [comment]
> Okay, so next time I fork a project I should name it differently?

Not the project, no. Here's the process I usually follow when preparing a patch:

~~~console
λ ~: git clone git@github.com:piliugin-anton/uQuik.git uquik && cd $_
λ uquik (master): git checkout -b topic-branch
Switched to a new branch 'topic-branch'
λ uquik (topic-branch): printf '\n' >> .prettierignore 
λ uquik (topic-branch): printf '\n' >> .env
λ uquik (topic-branch): git add .prettierignore .env
λ uquik (topic-branch): git commit
[topic-branch 1155331] Add trailing newlines to files missing them
 2 files changed, 2 insertions(+), 2 deletions(-)
~~~

Then I switch to GitHub and create the [actual fork](https://github.com/Cutlery-Drawer/uQuik):

Figure 1

Then in the terminal:

~~~console
λ uquik (topic-branch): git remote add fork git@github.com:Cutlery-Drawer/uQuik.git
λ uquik (topic-branch): git push -u fork topic-branch
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 16 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (4/4), 352 bytes | 352.00 KiB/s, done.
Total 4 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
remote: 
remote: Create a pull request for 'topic-branch' on GitHub by visiting:
remote:      https://github.com/Cutlery-Drawer/uQuik/pull/new/topic-branch
remote: 
To github.com:Cutlery-Drawer/uQuik.git
 * [new branch]      topic-branch -> topic-branch
branch 'topic-branch' set up to track 'fork/topic-branch'.
~~~

Finally, I switch back to GitHub, and see this:
Figure 2

**Comment by Alhadis** [comment]
> I could try to update the ESLint version

You might want to update my ESLint config as well, as I'm not sure it'll be compatible with newer ESLint versions. Latest version is [`v2.4.1`](https://github.com/Alhadis/JG/blob/a6f45f6ef21dc5026221ce84cbdfbc883221fbc0/eslint/package.json#L3).

**Comment by piliugin-anton** [comment]
> ```
> λ ~: git clone git@github.com:piliugin-anton/uQuik.git uquik && cd $_
> λ uquik (master): git checkout -b topic-branch
> Switched to a new branch 'topic-branch'
> λ uquik (topic-branch): printf '\n' >> .prettierignore 
> λ uquik (topic-branch): printf '\n' >> .env
> λ uquik (topic-branch): git add .prettierignore .env
> λ uquik (topic-branch): git commit
> [topic-branch 1155331] Add trailing newlines to files missing them
>  2 files changed, 2 insertions(+), 2 deletions(-)
> ```

Aha, so I just need to create a new branch and commit all changes to it, instead of master branch?

**Comment by Alhadis** [comment]
> Aha, so I just need to create a new branch and commit all changes to it, instead of master branch?

Yep! 😀

**Comment by piliugin-anton** [comment]
Okay, I had to remove your ESLint config, because of strange list of errors. So now it only uses TypeScript plugin

**Comment by Alhadis** [comment]
> Okay, I had to remove your ESLint config, because of strange list of errors

Wait, what weird errors? I added `"@alhadis/eslint-config": "^2.4.1"` to the `devDependencies` and it linted perfectly fine…

**Comment by piliugin-anton** [comment]
> > Okay, I had to remove your ESLint config, because of strange list of errors
> 
> Wait, what weird errors? I added `"@alhadis/eslint-config": "^2.4.1"` to the `devDependencies` and it linted perfectly fine…

Screenshot from 2022-09-05 03-49-58

**Comment by Alhadis** [comment]
That config is using ESLint v8.23.0. Are your other ESLint-using dependencies up-to-date? 🤔

**Comment by piliugin-anton** [comment]
> That config is using ESLint v8.23.0. Are your other ESLint-using dependencies up-to-date? thinking

Yes, I wrote you before, that I'll try to update ESLint. After update it shows me this errors.

**ESLint version 8.23.0**, **@alhadis/eslint-config version 2.4.1**

**.eslintrc.js**
```js
module.exports = {
	extends: '@alhadis',
  	parser: '@typescript-eslint/parser',
  	plugins: ['@typescript-eslint'],
  	overrides: [
    	{
      		files: [
        		"test/*"
      		],
      		globals: {
        		expect: true
      		}
    	}
  	]
};
```

**Comment by Alhadis** [comment]
No idea. Okay, yeah, screw it: ditch my config and do whatever. 😂 This is about to be your repository anyway, so I've no business pushing my linting preferences where they don't belong.

**Comment by Alhadis** [comment]
BTW, you'll need to delete your fork before I can complete the transfer. 😉 It's telling me `piliugin-anton/PanAndZoom` already exists. 😁

**Comment by piliugin-anton** [comment]
> BTW, you'll need to delete your fork before I can complete the transfer. wink It's telling me `piliugin-anton/PanAndZoom` already exists. grin

Done =)
Will you publish new versions in future? Will you be able to contribute if I will have a math problem?

**Comment by Alhadis** [comment]
> Will you publish new versions in future?

Well, no, that's your responsibility, my friend. 😉 What's your username at `npmjs.com`? Mine is, well, the same as it's always been for as long as I've used the Internet: [Alhadis](https://www.npmjs.com/~alhadis). 😁

> Will you be able to contribute if I will have a math problem?

Maths, no. Transform matrices, quite possibly. Keep in mind I'm actually [dyscalculic](https://dsf.net.au/learning-difficulties/dyscalculia/what-is-dyscalculia) and can really only understand numbers when I can see—visually—what they do when I fiddle with them. That being said, I'll help out as much as I can, but there's a pretty good chance you'll end up teaching _me_ about math instead of the other way around. 🤣

**Comment by piliugin-anton** [comment]
> What's your username at `npmjs.com`?

anton.piliugin (I just updated the photo, so it'll appear soon)

> That being said, I'll help out as much as I can, but there's a pretty good chance you'll end up teaching _me_ about math instead of the other way around. rofl

That would be an interesting outcome of our collaboration =)

**Comment by Alhadis** [comment]
Righto, invite sent. 👍 Keep in mind I won't be doing any further work on this repository, as I've too many damn projects already (and more on the way…), so I thank you for being willing to take this one off my hands. 🙇

**Comment by piliugin-anton** [comment]
> Righto, invite sent. +1 Keep in mind I won't be doing any further work on this repository, as I've too many damn projects already (and more on the way…), so I thank you for being willing to take this one off my hands. bow

Accepted invite. I'm happy to contribute to your project, in my plans to create some kind of highly performant framework for working with canvas 2D (I will benchmark your solution with the same but in WASM)

**Comment by Alhadis** [comment]
> I will benchmark your solution with the same but in WASM

LOL, WebAssembly has [SIMD operators](https://webassembly.github.io/simd/core/appendix/index-instructions.html), dude. 🤣 Higher-level dynamic languages don't stand a snowball's chance in hell.

**Comment by piliugin-anton** [comment]
> > I will benchmark your solution with the same but in WASM
> 
> LOL, WebAssembly has [SIMD operators](https://webassembly.github.io/simd/core/appendix/index-instructions.html), dude. rofl Higher-level dynamic languages don't stand a snowball's chance in hell.

Looks like Safari still has no SIMD support

**Comment by Alhadis** [comment]
And macOS's Terminal.app doesn't support [OSC 8 hyperlinks](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda#) either. Apple are always lagging behind when it comes to implementing shiny new goodies…

BTW, I literally just learned that it's possible to do [this](https://github.com/tc39/ecma262/issues/543) in ECMAScript. What the actual _fuck_:

~~~js
class FuckAll extends null {
	constructor(){
		console.log(this);
	}
}
new FuckAll();
~~~

And predictably enough, this raises the usual `ReferenceError`:

> Uncaught ReferenceError: Must call super constructor in derived class before accessing 'this' or returning from derived constructor

But calling `super()` is useless too:

> Uncaught TypeError: Super constructor null of FuckAll is not a constructor
    at new FuckAll (<anonymous>:4:3)
    at <anonymous>:8:1

I haven't felt this confused over a JS feature since discovering `typeof document.all`.

**Comment by piliugin-anton** [comment]
You can also work with arrays using negative indexes: https://youtu.be/pFEdFfB2ewE?t=38
I have to go to sleep, tomorrow I will release a new version to NPM

**Comment by Alhadis** [comment]
> You can also work with arrays using negative indexes:

That's nothing. Watch this:

~~~js
const empty = [];
Array.prototype.push("lolwut");
empty[0] === "lolwut"
~~~

>  I have to go to sleep, tomorrow I will release a new version to NPM

All good mate. It's 10:40 in the morning here, haha. 🇦🇺 Sleep well wherever you live. 👍

**Comment by piliugin-anton** [comment]
Finally, released a new version. Added a few tests for update callbacks.
